File: hardening/iso_stamp_oracle.php
<?php
// Independent-runtime oracle: parses one ISO 8601 timestamp with PHP's DateTimeImmutable and prints four fields.
// Usage: php iso_stamp_oracle.php '<timestamp>'
$stamp = new DateTimeImmutable($argv[1]);
echo $stamp->format('Y-m-d\TH:i:s'), "\n";
echo $stamp->format('u'), "\n";
echo $stamp->getOffset(), "\n";
echo $stamp->format('P'), "\n";

File: hardening/iso_stamp_oracle.rb
# Independent-runtime oracle: parses one ISO 8601 timestamp with Ruby's Time.iso8601 and prints four fields.
# Usage: ruby iso_stamp_oracle.rb '<timestamp>'
require 'time'
STDOUT.set_encoding('UTF-8')
stamp = Time.iso8601(ARGV[0].dup.force_encoding('UTF-8'))
puts stamp.strftime('%Y-%m-%dT%H:%M:%S')
puts stamp.usec
puts stamp.nsec
puts stamp.utc_offset
